pkg/health/heartbeat/heartbeat.go

```diff
@@ -0,0 +1,277 @@
+	renewedLease.Spec.HolderIdentity = &actor
+	renewedTime := time.Now()
+	renewedLease.Spec.RenewTime = &metav1.MicroTime{Time: renewedTime}
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		return k8sClientset.Patch(context.Background(), renewedLease, client.MergeFrom(fullSnapLease))
+	})
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to update full snapshot lease: %v", err),
+		}
+	}
+	logger.Info("Renewed full snapshot lease with revision ", actor, " at time ", renewedTime)
+
+	return nil
+}
+
+// UpdateDeltaSnapshotLeases updates the holderIdentity field of the delta snapshot lease with the total number or revisions stored in delta snapshots since the last full snapshot was taken
+func UpdateDeltaSnapshotLeases(logger *logrus.Entry, fullSnapshot *brtypes.Snapshot, PrevDeltaSnapshots brtypes.SnapList, k8sClientset client.Client) error {
+
+	namespace, err := GetEnvVarOrError("POD_NAMESPACE")
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Pod namespace env var not present: %v", err),
+		}
+	}
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to create k8s clientset: %v", err),
+		}
+	}
+
+	deltaSnapLease := &v1.Lease{}
+	err = k8sClientset.Get(context.TODO(), client.ObjectKey{
+		Namespace: namespace,
+		Name:      brtypes.DeltaSnapshotLeaseName,
+	}, deltaSnapLease)
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to fetch delta snapshot lease: %v", err),
+		}
+	}
+	renewedLease := deltaSnapLease.DeepCopy()
+	actor := "0"
+	if len(PrevDeltaSnapshots) > 0 {
+		fullSnapshotRevisions := int64(0)
+		if fullSnapshot != nil {
+			fullSnapshotRevisions = int64(fullSnapshot.LastRevision)
+		}
+		deltaSnap := PrevDeltaSnapshots[len(PrevDeltaSnapshots)-1]
+		deltaSnapshotRevision := deltaSnap.LastRevision - int64(fullSnapshotRevisions)
+		actor = strconv.FormatInt(deltaSnapshotRevision, 10)
+	}
+	renewedLease.Spec.HolderIdentity = &actor
+	renewedTime := time.Now()
+	renewedLease.Spec.RenewTime = &metav1.MicroTime{Time: renewedTime}
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		return k8sClientset.Patch(context.Background(), renewedLease, client.MergeFrom(deltaSnapLease))
+	})
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to update delta snapshot lease: %v", err),
+		}
+	}
+	logger.Info("Renewed delta snapshot lease with revision ", actor, " at time ", renewedTime)
+
+	return nil
+}
+
+// RenewMemberLeasePeriodically has a timer and will periodically call RenewMemberLeases to renew the member lease until stopped
+func RenewMemberLeasePeriodically(ctx context.Context, hconfig *brtypes.HealthConfig, logger *logrus.Entry, etcdConfig *etcdutil.EtcdConnectionConfig) error {
+
+	hb, err := NewHeartbeat(hconfig, logger, etcdConfig)
+	if err != nil {
+		hb.logger.Warnf("Could not initialize new heartbeat: %v", err)
```

Comment:
Why do we only log but continue when there was an error when creating the Heartbeat instance?

Reply:
Thanks for catching this
Updated to return an error if heartbeat could not be initialized